cmd/snap/cmd_buy_test.go

```diff
@@ -194,39 +231,293 @@ func (s *SnapSuite) TestBuySnap(c *check.C) {
+`)
+	c.Check(s.Stderr(), check.Equals, "")
+	c.Check(findCount, check.Equals, 1)
+	c.Check(buyMethodsCount, check.Equals, 1)
 	c.Check(postCount, check.Equals, 1)
 }
 
+const buyNoPaymentMethodsJson = `
+{
+  "type": "sync",
+  "status-code": 200,
+  "status": "OK",
+  "result": {
+    "allows-automatic-payment": false,
+    "methods": []
+  },
+  "sources": [
+    "store"
+  ],
+  "suggested-currency": "GBP"
+}
+`
+
+func (s *SnapSuite) TestBuySnapFailsNoPaymentMethods(c *check.C) {
+	findCount := 0
+	buyMethodsCount := 0
+	s.RedirectClientToTestServer(func(w http.ResponseWriter, r *http.Request) {
+		switch r.Method {
+		case "GET":
+			switch r.URL.Path {
+			case "/v2/find":
+				q := r.URL.Query()
+				c.Check(q.Get("q"), check.Equals, "name:hello")
+				fmt.Fprintln(w, buySnapFindJson)
+				findCount++
+			case "/v2/buy/methods":
+				fmt.Fprintln(w, buyNoPaymentMethodsJson)
+				buyMethodsCount++
+			default:
+				c.Fatalf("unexpected URL %q", r.URL.Path)
+			}
+		default:
+			c.Fatalf("unexpected HTTP method %q", r.Method)
+		}
+	})
+
+	rest, err := snap.Parser().ParseArgs([]string{"buy", "hello"})
+	c.Assert(err, check.NotNil)
+	c.Check(err.Error(), check.Equals, `cannot buy snap "hello": no payment methods registered`)
+	c.Check(rest, check.DeepEquals, []string{"hello"})
+	c.Check(s.Stdout(), check.Equals, "")
+	c.Check(s.Stderr(), check.Equals, "")
+	c.Check(findCount, check.Equals, 1)
+	c.Check(buyMethodsCount, check.Equals, 1)
+}
+
+func (s *SnapSuite) TestBuySnapFailsInvalidMethodID(c *check.C) {
+	findCount := 0
+	buyMethodsCount := 0
+	s.RedirectClientToTestServer(func(w http.ResponseWriter, r *http.Request) {
+		switch r.Method {
+		case "GET":
+			switch r.URL.Path {
+			case "/v2/find":
+				q := r.URL.Query()
+				c.Check(q.Get("q"), check.Equals, "name:hello")
+				fmt.Fprintln(w, buySnapFindJson)
+				findCount++
+			case "/v2/buy/methods":
+				fmt.Fprintln(w, buyMethodsSelectPaymentMethodJson)
+				buyMethodsCount++
+			default:
+				c.Fatalf("unexpected URL %q", r.URL.Path)
+			}
+		default:
+			c.Fatalf("unexpected HTTP method %q", r.Method)
+		}
+	})
+
+	// Type an invalid number
+	fmt.Fprint(s.stdin, "abc\n")
+
+	rest, err := snap.Parser().ParseArgs([]string{"buy", "hello"})
+	c.Assert(err, check.NotNil)
+	c.Check(err.Error(), check.Equals, `cannot buy snap "hello": invalid payment method selection "abc"`)
+	c.Check(rest, check.DeepEquals, []string{"hello"})
+	c.Check(s.Stdout(), check.Equals, `  Selection  Description
+  1          **** **** **** 1111 (exp 23/2020)
+  2          **** **** **** 2222 (exp 23/2025)
+  3          PayPal
+Select payment method: `)
+	c.Check(s.Stderr(), check.Equals, "")
+	c.Check(findCount, check.Equals, 1)
+	c.Check(buyMethodsCount, check.Equals, 1)
+}
+
+func (s *SnapSuite) TestBuySnapFailsOutOfRangeMethodID(c *check.C) {
+	findCount := 0
+	buyMethodsCount := 0
+	s.RedirectClientToTestServer(func(w http.ResponseWriter, r *http.Request) {
+		switch r.Method {
```

Comment:
The function inside `RedirectClientToTestServer` looks the same for most (all?) of the tests, expect the returned json. It seems it would make sense to extract them into a common helper that just gets the json it should return?